File: ChatBotFAQ/app/services/retriever.py
"""Retrieve the passages that should ground an answer.

The search is two-stage. FAISS finds a wide band of candidates on dense
similarity, then those candidates are rescored against the query's lexical
weights and only the best few survive. Both scores come from the *same* BGE-M3
forward pass over the question, so the second stage costs no model call at all —
which is why it is done here rather than with a cross-encoder. On this CPU a
cross-encoder over 20 candidates was measured at 3.6-32s depending on size;
this is a dictionary intersection.

What the lexical half buys: dense embeddings smooth away exact terms. The page
`/courses/ai-blockchain-fintech-for-beginners` never surfaced for "khóa học nào
phù hợp với người mới bắt đầu" even though its slug says *beginners*, because
the phrase appears in its title and not its prose. Lexical weights match the
token itself, so it can.

Three filters then apply, each because of a measured property of this corpus:

  * **Weak matches are dropped.** BGE-M3 cosine scores run about 0.60-0.78 for
    passages that answer the question; below ~0.35 the hits are unrelated. FAISS
    returns k results however poor the match, so without a floor an off-topic
    question still fills the prompt with confident-looking noise.

  * **One page cannot take every slot.** A long article contributes many chunks
    that all score alike, crowding out the short page that holds the answer.
    Capping per URL keeps the results spread across sources. Applied *after*
    reranking, so the cap trims the final ranking rather than the raw one.

  * **Over-fetch, then filter.** Both of the above remove hits, so the search
    asks for more than it needs and trims back.
"""
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from app.services.Data_embedding import DataEmbedding
from app.services.vector_store import VectorStore

DEFAULT_INDEX_DIR = Path(__file__).resolve().parents[2] / "data" / "faiss_index"

# Cosine floor. Measured on this index: answering passages score 0.60-0.78,
# unrelated ones fall well below this.
DEFAULT_MIN_SCORE = 0.35

# No single page may occupy more than this many of the final slots.
DEFAULT_MAX_PER_URL = 3

# How many candidates the dense stage hands to reranking. Wide enough to give
# the lexical half something to reorder, narrow enough that the score floor has
# usually already cut it below this anyway.
DEFAULT_CANDIDATES = 20

# Hybrid weights. Dense leads because it carries the meaning; lexical corrects
# it. Tune with scripts/tune_hybrid.py rather than by intuition.
DENSE_WEIGHT = float(os.environ.get("CHATBOT_DENSE_WEIGHT", "0.7"))
SPARSE_WEIGHT = float(os.environ.get("CHATBOT_SPARSE_WEIGHT", "0.3"))

# Words a user says when they mean "this institute" without naming it — this
# bot only ever serves one organization, so there is no ambiguity to resolve,
# just a name missing from the query that the corpus's own pages spell out in
# full. Kept short and reviewed by hand rather than inferred, since a wrong
# entry here silently pollutes every query that contains it.
INSTITUTE_ALIASES = ["viện", "trường"]
INSTITUTE_FULL_NAME = "Viện Công nghệ và Kinh tế số BK Fintech"


def expand_self_reference(question: str) -> str | None:
    """A second query variant with the institute's full name appended, for
    questions that refer to it only as "viện"/"trường".

    Returns None (rather than the question unchanged) when no alias is
    present, so callers can tell "nothing to add" apart from "added and it's
    a no-op" without a second check.
    """
    q = question.lower()
    if INSTITUTE_FULL_NAME.lower() in q:
        return None
    if any(alias in q for alias in INSTITUTE_ALIASES):
        return f"{question} {INSTITUTE_FULL_NAME}"
    return None


def _rescale(values: list[float]) -> list[float]:
    """Stretch a list onto 0..1. A flat list carries no ranking signal, so it
    becomes all zeros rather than an arbitrary order."""
    low, high = min(values), max(values)
    if high - low < 1e-9:
        return [0.0] * len(values)
    return [(v - low) / (high - low) for v in values]


class Retriever:
    def __init__(
        self,
        index_dir: Path = DEFAULT_INDEX_DIR,
        embedder: DataEmbedding = None,
        translator=None,
    ):
        self.store = VectorStore.load(index_dir)
        # Loading the model costs seconds and ~2GB, so it is created once and
        # shared; pass one in to reuse an existing instance.
        self.embedder = embedder or DataEmbedding()
        # Optional: pass a QueryTranslator to search English passages with an
        # English query. Left None by scripts that must not make API calls.
        self.translator = translator

    def query_for(self, question: str) -> str:
        """The text actually embedded — English if the question was translated.

        Public so callers can show it: a user seeing a Vietnamese question
        return English-looking matches deserves to know why, and it is the first
        thing to check when retrieval goes wrong.
        """
        return self.translator.to_english(question) if self.translator else question

    def search(
        self,
        question: str,
        top_k: int = 7,
        min_score: float = DEFAULT_MIN_SCORE,
        max_per_url: int = DEFAULT_MAX_PER_URL,
        candidates: int = DEFAULT_CANDIDATES,
        rerank: bool = True,
    ) -> list[dict]:
        # Search with the question and, when it was translated, with the English
        # version too. The local translator mangles domain terms often enough
        # ("Viện trưởng" -> "the Chief") that replacing the query outright would
        # sometimes lose a match the original would have found. Keeping both and
        # taking the better score per chunk can only add, never subtract.
        queries = [question]
        expanded = expand_self_reference(question)
        if expanded:
            queries.append(expanded)
        # Translate the expanded form when there is one: "viện" alone commonly
        # mistranslates to "hospital" (the more frequent sense in general text),
        # but with the institute's full name attached the translator has an
        # anchor and resolves it correctly.
        english = self.query_for(expanded or question)
        if english != question:
            queries.append(english)

        hits: dict[int, dict] = {}
        lexicals = []
        for text in queries:
            query_dense, query_lexical = self.embedder.embed_query(text)
            lexicals.append(query_lexical)
            for hit in self.store.search(query_dense, top_k=max(candidates, top_k)):
                if hit["score"] < min_score:
                    continue
                seen = hits.get(hit["index"])
                if seen is None or hit["score"] > seen["score"]:
                    hits[hit["index"]] = hit
        hits = sorted(hits.values(), key=lambda h: h["score"], reverse=True)[:candidates]

        if rerank and any(lexicals) and len(hits) > 1:
            for hit in hits:
                doc_lexical = self.store.lexical_for(hit["index"])
                hit["dense_score"] = hit["score"]
                # Best of both queries again: the Vietnamese one shares tokens
                # with the Vietnamese book chunks, the English one with the
                # English site pages. Neither covers both halves of the corpus.
                hit["lexical_score"] = max(
                    (self.embedder.lexical_score(q, doc_lexical) for q in lexicals),
                    default=0.0,
                ) if doc_lexical else 0.0
            # The two scores are not on the same scale — dense cosine sits
            # around 0.4-0.8 while a lexical match measures ~0.0-0.3, so adding
            # them raw lets dense decide everything and the weights mean
            # nothing. Both are stretched across the candidate set first, which
            # makes the weights a real ratio between the two signals.
            dense = _rescale([h["dense_score"] for h in hits])
            lexical = _rescale([h["lexical_score"] for h in hits])
            for hit, d, s in zip(hits, dense, lexical):
                hit["score"] = DENSE_WEIGHT * d + SPARSE_WEIGHT * s
            hits.sort(key=lambda h: h["score"], reverse=True)

        kept: list[dict] = []
        per_url: dict[str, int] = {}
        for hit in hits:
            url = hit.get("url") or ""
            if per_url.get(url, 0) >= max_per_url:
                continue
            per_url[url] = per_url.get(url, 0) + 1
            kept.append(hit)
            if len(kept) >= top_k:
                break
        return kept

File: src/lib/chatbot/retriever.ts
/**
 * Retrieve the passages that should ground an answer. Direct port of
 * retriever.py's two-stage search — see that file's module docstring for the
 * full reasoning; kept here only where the JS port changes something.
 *
 * Stage 0 (query building): the question is turned into two search texts —
 * bản tiếng Việt (khôi phục dấu nếu cần, xem diacritics.ts) và một truy vấn
 * tiếng Anh độc lập do LLM dựng lại (xem queryRewriter.ts). Bước rewrite này
 * thay cho cặp expand_self_reference + bản dịch Marian mà retriever.py dùng;
 * đừng đi tìm hàm tương ứng bên đó.
 * Stage 1 (dense): VectorStore.search() over the whole corpus.
 * Stage 2 (lexical rerank): re-score the dense candidates against the
 * query's lexical weights and blend the two — see embedding.ts's top comment
 * for how the lexical signal here differs from the Python version's learned
 * BGE-M3 sparse output.
 *
 * Three filters apply after that, same as retriever.py and same reasons:
 *   - weak matches dropped below DEFAULT_MIN_SCORE
 *   - no single URL takes more than DEFAULT_MAX_PER_URL of the final slots
 *   - candidates are over-fetched so both filters have room to trim
 *
 * `options.history` cũng KHÔNG phải phần port: retriever.py không nhận history
 * và search() bên đó không có tham số tương ứng. Ở đây history chỉ đi đúng một
 * chỗ — bước rewrite, để giải đại từ. Nó từng nuôi thêm một biến thể truy vấn
 * thứ ba (query ghép ngữ cảnh); biến thể đó đã bị xoá, xem contextQuery.ts.
 */
import {
  DEFAULT_CANDIDATES,
  DEFAULT_MAX_PER_URL,
  DEFAULT_MIN_SCORE,
  DEFAULT_TOP_K,
  DENSE_WEIGHT,
  SPARSE_WEIGHT,
} from "./config";
import {
  embedQuery,
  lexicalScore,
  loadEmbedder,
  type LexicalWeights,
} from "./embedding";
import { getStore, VectorStore, type SearchHit } from "./vectorStore";
import { rewriteQuery } from "./queryRewriter";
import { needsRestoration, restoreQuestion } from "./diacritics";
// `import type`, not a value import, and it has to stay that way: TS erases it
// completely at compile time, so the mongodb driver chatHistory.ts pulls in
// stays out of this module's runtime import graph. Turn it into a plain import
// and every file that touches the retriever drags Mongo along with it, with
// nothing at the call site to explain why the bundle grew.
import type { ChatMessage } from "./chatHistory";

function rescale(values: number[]): number[] {
  if (values.length === 0) return values;
  const low = Math.min(...values);
  const high = Math.max(...values);
  if (high - low < 1e-9) return values.map(() => 0);
  return values.map((v) => (v - low) / (high - low));
}

interface RankedHit extends SearchHit {
  dense_score?: number;
  lexical_score?: number;
}

export interface RetrievalChunk extends SearchHit {
  dense_score?: number;
  lexical_score?: number;
}

export interface RetrievalResult {
  chunks: RetrievalChunk[];
  /** Truy vấn tiếng Anh đã đem đi nhúng, hoặc chính câu hỏi nếu bước rewrite bị
   * bỏ qua. Trả ra ngoài chứ không để caller tự dựng lại: hai đường dựng query
   * độc lập là đúng cái bug vừa sửa ở đây (queryFor() và search() gọi bước dịch
   * hai lần với chuỗi hơi khác nhau, miss cache cả hai lần). Public vì người
   * dùng thấy hỏi tiếng Việt mà khớp tài liệu tiếng Anh thì xứng đáng biết lý
   * do — cùng mục đích như query_for() bên retriever.py. */
  searchQuery: string;
}

export class Retriever {
  /** Store được ghim cứng do caller truyền vào — chỉ các script đo đạc muốn
   *  một bản bất biến trong suốt lượt chạy mới dùng tới.
   *
   *  null (mặc định, và là đường mà production đi) = dùng store CHUNG của tiến
   *  trình, và store đó tự nạp lại khi store.json trên đĩa đổi. Đây là chỗ bỏ
   *  giả định cũ "store không bao giờ đổi trong đời tiến trình" — giả định đã
   *  buộc phải restart app sau mỗi lần cập nhật dữ liệu. Xem getStore(). */
  private readonly pinned: VectorStore | null;

  constructor(store?: VectorStore) {
    this.pinned = store ?? null;
  }

  private async store(): Promise<VectorStore> {
    return this.pinned ?? getStore();
  }

  /**
   * Rebuild chunks the caller already identified, in the order given — no
   * embedding, no search. Used by the follow-up path: when a question refers
   * back ("người thứ 2") and retrieval on its own found nothing anchored, the
   * passages that answered the *previous* turn are by construction where the
   * referent came from.
   *
   * The scores come from that previous turn and are carried, not recomputed:
   * these passages were never scored against the current question, and inventing
   * a number here would put a fiction in the source list. Ids missing from the
   * index are skipped — the index can be rebuilt between turns.
   */
  async chunksByIds(
    entries: { id: string; score: number }[]
  ): Promise<RetrievalChunk[]> {
    const store = await this.store();
    const out: RetrievalChunk[] = [];
    for (const { id, score } of entries) {
      const index = store.chunkIdToIndex(id);
      if (index === undefined) continue;
      out.push(store.hitAt(index, score));
    }
    return out;
  }

  async search(
    question: string,
    options: {
      topK?: number;
      minScore?: number;
      maxPerUrl?: number;
      candidates?: number;
      rerank?: boolean;
      /** This session's earlier turns, oldest first. Chỉ dùng cho bước rewrite
       * (giải đại từ theo lượt trước). Bỏ qua thì rewrite vẫn chạy, chỉ là
       * không có gì để giải — và với câu hỏi tiếng Anh ở lượt đầu thì bước đó
       * bị bỏ hẳn, xem rewriteQuery(). */
      history?: ChatMessage[];
    } = {}
  ): Promise<RetrievalResult> {
    const topK = options.topK ?? DEFAULT_TOP_K;
    const minScore = options.minScore ?? DEFAULT_MIN_SCORE;
    const maxPerUrl = options.maxPerUrl ?? DEFAULT_MAX_PER_URL;
    const candidates = options.candidates ?? DEFAULT_CANDIDATES;
    const rerank = options.rerank ?? true;

    // Song song hoá, và đây là chỗ giấu được gần hết chi phí của bước rewrite:
    // trên cold start, call LLM chạy trọn trong bóng của ~5s nạp BGE-M3 nên
    // gần như miễn phí; trên instance đã nóng nó lộ ra ~1s — nhưng đó là ~1s
    // thay cho ~20s của hai lần chạy model dịch trước đây. loadEmbedder() được
    // gọi ở đây chỉ để hâm nóng song song: nó đã cache trên globalThis nên là
    // no-op nếu instrumentation đã nạp, và kết quả không dùng tới ở đây.
    //
    // rewriteQuery() nhận câu hỏi THÔ chứ không phải bản đã khôi phục dấu: nó
    // tự xử lý được câu không dấu (xem REWRITE_SYSTEM_PROMPT), và nhận câu thô
    // là thứ cho phép nó chạy song song với chính bước khôi phục dấu.
    const [store, rewritten, viterbi] = await Promise.all([
      this.store(),
      rewriteQuery(question, options.history ?? []),
      // Khôi phục dấu bằng Viterbi. Vẫn chạy song song vô điều kiện dù bước
      // rewrite thường cho kết quả tốt hơn, vì nó là NGUỒN DUY NHẤT không phụ
      // thuộc mạng: ~0.5ms, không bao giờ hỏng. Câu vốn đã có dấu đi thẳng qua
      // không bị đụng vào (xem needsRestoration).
      restoreQuestion(question),
      loadEmbedder(),
    ]);

    // Dấu do LLM khôi phục thắng Viterbi khi có. Đo trên 29 câu tiếng Việt của
    // qa-cases (bỏ dấu rồi khôi phục, so với chính câu gốc): Viterbi 51.7% đúng
    // cả câu, LLM 93.1%. Quan trọng hơn con số tổng là KIỂU sai của Viterbi —
    // nó biến "khóa học" thành "khoa học" ở 4/29 câu, mà corpus này có cả trang
    // khóa học lẫn trang nghiên cứu khoa học, nên đó là lái truy vấn sang đúng
    // collection sai. Bảng bigram không cứu được: "khoa học" phổ biến hơn hẳn
    // trong chính corpus, tức thống kê corpus phản lại người dùng ở đúng chỗ đó.
    //
    // Chỉ áp dụng khi needsRestoration() đúng — tức đúng lúc Viterbi vốn sẽ
    // chạy. Câu người dùng đã tự gõ dấu thì không ai đoán hộ, giữ nguyên lập
    // trường đã ghi trong diacritics.ts. (Câu gõ lẫn lộn vẫn nằm ngoài, y như
    // trước.) Dòng VI đã qua kiểm tra chuỗi từ ở parseRewrite(), nên tới đây nó
    // chắc chắn là cùng những chữ đó theo cùng thứ tự đó.
    const restored =
      rewritten.vietnamese && needsRestoration(question)
        ? rewritten.vietnamese
        : viterbi;

    // Vẫn tìm bằng CẢ câu tiếng Việt gốc, hai lý do và cả hai đều còn nguyên
    // giá trị sau khi đổi từ dịch máy sang rewrite bằng LLM:
    //   1. một bản viết lại sai chỉ được phép THÊM match, không bao giờ được
    //      thay câu đúng bằng câu sai — hits gộp theo điểm tốt nhất mỗi chunk,
    //      nên một biến thể vô ích tốn đúng một lượt nhúng chứ không đẩy được
    //      query tốt ra ngoài;
    //   2. nửa lexical của hybrid search cần token tiếng Việt. Chunk tiếng Việt
    //      trong corpus không khớp gì với query tiếng Anh, nên bỏ câu gốc là tự
    //      cắt một nửa tín hiệu TF-IDF.
    const queries = [restored];
    if (rewritten.english.toLowerCase() !== restored.toLowerCase()) {
      queries.push(rewritten.english);
    }

    const hitsByIndex = new Map<number, RankedHit>();
    const lexicals: LexicalWeights[] = [];

    for (const query of queries) {
      const { dense, lexical } = await embedQuery(query, store.idf);
      lexicals.push(lexical);
      for (const hit of store.search(dense, Math.max(candidates, topK))) {
        // minScore trả lời "chunk này có đủ gần để làm ứng viên không" — thuộc
        // tính của phép khớp, không phụ thuộc biến thể nào tìm ra nó.
        if (hit.score < minScore) continue;
        const seen = hitsByIndex.get(hit.index);
        if (!seen || hit.score > seen.score) hitsByIndex.set(hit.index, hit);
      }
    }

    const ranked = Array.from(hitsByIndex.values()).sort(
      (a, b) => b.score - a.score
    );
    const hits = ranked.slice(0, candidates);

    const anyLexical = lexicals.some((l) => l.size > 0);
    if (rerank && anyLexical && hits.length > 1) {
      for (const hit of hits) {
        const docLexical = store.lexicalFor(hit.index);
        hit.dense_score = hit.score;
        // Best of both queries, same as retriever.py: the Vietnamese query
        // shares tokens with Vietnamese content, the rewritten English one with
        // English pages — neither alone covers both halves of the corpus.
        hit.lexical_score = docLexical.size
          ? Math.max(...lexicals.map((q) => lexicalScore(q, docLexical)), 0)
          : 0;
      }
      const dense = rescale(hits.map((h) => h.dense_score ?? h.score));
      const lexical = rescale(hits.map((h) => h.lexical_score ?? 0));
      hits.forEach((hit, i) => {
        hit.score = DENSE_WEIGHT * dense[i] + SPARSE_WEIGHT * lexical[i];
      });
      hits.sort((a, b) => b.score - a.score);
    }

    const kept: RetrievalChunk[] = [];
    const perUrl = new Map<string, number>();
    for (const hit of hits) {
      const url = hit.url || "";
      const count = perUrl.get(url) || 0;
      if (count >= maxPerUrl) continue;
      perUrl.set(url, count + 1);
      kept.push(hit);
      if (kept.length >= topK) break;
    }
    return { chunks: kept, searchQuery: rewritten.english };
  }
}
